infrastructure/src/main/java/org/corfudb/infrastructure/management/failuredetector/HealingAgent.java

```diff
@@ -0,0 +1,154 @@
+package org.corfudb.infrastructure.management.failuredetector;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.RemoteMonitoringService.DetectorTask;
+import org.corfudb.infrastructure.management.ClusterAdvisor;
+import org.corfudb.infrastructure.management.PollReport;
+import org.corfudb.infrastructure.management.ReconfigurationEventHandler;
+import org.corfudb.infrastructure.redundancy.RedundancyCalculator;
+import org.corfudb.protocols.wireprotocol.failuredetector.FailureDetectorMetrics;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeRank;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.view.Layout;
+import org.corfudb.util.concurrent.SingletonResource;
+
+import java.time.Duration;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+
+import static org.corfudb.infrastructure.RemoteMonitoringService.DetectorTask.SKIPPED;
+
+/**
+ * Healing algorithm
+ */
+@Slf4j
+@Builder
+public class HealingAgent {
+    /**
+     * Duration in which the restore redundancy and merge segments workflow status is queried.
+     */
+    @Builder.Default
+    private final Duration mergeSegmentsRetryQueryTimeout = Duration.ofSeconds(1);
+
+    @NonNull
+    private final ClusterAdvisor advisor;
+
+    @NonNull
+    private final String localEndpoint;
+    /**
+     * Future which is reset every time a new task to mergeSegments is launched.
+     * This is to avoid multiple mergeSegments requests.
+     */
+    @Builder.Default
+    private volatile CompletableFuture<Boolean> mergeSegmentsTask = CompletableFuture.completedFuture(true);
+
+    private final CompletableFuture<DetectorTask> skippedTask = CompletableFuture.completedFuture(SKIPPED);
+
+    @NonNull
+    private final SingletonResource<CorfuRuntime> runtimeSingleton;
+
+    @NonNull
+    private final ExecutorService failureDetectorWorker;
+
+    @NonNull
+    private final FailureDetectorDataStore dataStore;
+
+    /**
+     * Handle healed node.
+     * Cluster advisor provides healed node based on current cluster state if healed node found then
+     * save healed node in the history and send a message with the detected healed node
+     * to the relevant management server.
+     *
+     * @param pollReport poll report
+     * @param layout     current layout
+     */
+    public CompletableFuture<DetectorTask> detectHealing(PollReport pollReport, Layout layout) {
```

Comment:
Just a suggestion - Maybe making the method name detectAndHandleHealing would be more self explanatory.

Reply:
Done in part 3. https://github.com/CorfuDB/CorfuDB/pull/3112/commits/8bba9f9245fc1226753a04af0aa3c799adfc0ea4